node/src/Pos/Lrc/DB/Leaders.hs

```diff
@@ -44,13 +63,69 @@ prepareLrcLeaders ::
        , HasConfiguration
        )
     => m ()
-prepareLrcLeaders =
-    whenNothingM_ (getLeaders 0) $
-        putLeaders 0 genesisLeaders
+prepareLrcLeaders = do
+    -- Initialization flag was added with CSE-240.
+    unlessM isLrcDbInitialized $ do
+        hasLeadersForEpoch0 <- isJust <$> getLeadersForEpoch 0
+        if not hasLeadersForEpoch0 then
+            -- The node is not initialized at all. Only need to put leaders
+            -- for the first epoch.
+            putLeadersForEpoch 0 genesisLeaders
+        else
+            -- The node was initialized before CSE-240.
+            -- Need to migrate data for all epochs.
+            initLeaders 0
+  where
+    initLeaders :: MonadDB m => EpochIndex -> m ()
+    initLeaders i = do
+        whenNothingM_ (getLeader $ SlotId i minBound) $ do
+            maybeLeaders <- getLeadersForEpoch i
+            case maybeLeaders of
+                Just leaders -> do
+                    putBatchBi $ putLeadersForEpochSeparatelyOps i leaders
+                    initLeaders (i + 1)
+                Nothing -> pure ()
+
+isLrcDbInitialized :: MonadDB m => m Bool
+isLrcDbInitialized = dbHasKey lrcDbInitFlag
+
+putInitFlag :: MonadDB m => m ()
+putInitFlag = putBi lrcDbInitFlag ()
 
 ----------------------------------------------------------------------------
 -- Keys
 ----------------------------------------------------------------------------
 
-leadersKey :: EpochIndex -> ByteString
-leadersKey = mappend "l/" . serialize'
+leadersForEpochKey :: EpochIndex -> ByteString
+leadersForEpochKey = mappend "l/" . serialize'
+
+leaderKey :: HasProtocolConstants => SlotId -> ByteString
+leaderKey = mappend "ls/" . serialize' . flattenSlotId
+
+lrcDbInitFlag :: ByteString
+lrcDbInitFlag = "linit/"
```

Comment:
It should probably be renamed, because LRC DB includes not only leaders, but also other stuff. This particular file covers leaders part and this flag covers only leaders.
Name suggestions: `lrcDbLeadersInitFlag`, `leadersDbInitFlag`, …